packages/ui/src/components/controls/NumberInput.tsx

```diff
@@ -0,0 +1,126 @@
+import clsx from 'clsx';
+import type {JSX} from 'preact';
+import {useCallback, useEffect, useRef, useState} from 'preact/hooks';
+import {useDocumentEvent} from '../../hooks';
+import {MouseButton, clamp} from '../../utils';
+import styles from './Controls.module.scss';
+
+type NumberInputProps = Omit<
+  JSX.HTMLAttributes<HTMLInputElement>,
+  'min' | 'max' | 'step' | 'value' | 'onChange' | 'onChangeCapture'
+> & {
+  min?: number;
+  max?: number;
+  step?: number;
+  value: number;
+  onChange?: (value: number) => void;
+  onChangeCapture?: (value: number) => void;
+};
+
+export function NumberInput({
+  min = -Infinity,
+  max = Infinity,
+  step = 1,
+  value = 0,
+  onChange,
+  onChangeCapture,
+  ...props
+}: NumberInputProps) {
+  const handleChange = onChangeCapture ? onChange : undefined;
+  const handleChangeCapture = onChangeCapture ?? onChange;
```

Comment:
I don't think this is necessary here.
The `Input` component corrects the discrepancy between Preact's `onChange` and the browser's `onchange` by using `onChangeCapture` and since it's a generic component I left the possibility to not use the correction.
This is a pretty concrete component that has its own custom `onChange` callback so I think we can just dispatch it during `onChangeCapture` as that's how it's intended to work.

Reply:
So you mean we get rid of `onChangeCapture` entirely and use `onChange` instead? Would the `onChange` callback still be called for every step or, like `onChangeCapture` before, only when the pointer lock is released?

Reply:
The public interface should have an `onChange` property that would be called during `onChangeCapture` on the native input element.